### ISSUE TITLE
Storage error (only) when running in development mode

### ISSUE THREAD

**Issue by hugetim** [anvilistas/anvil-extras#106]
When clicking the "Run" button to test an Anvil app in development mode (but not when running the app via its url), I get the following error when trying to import `local_storage`:
```
ExternalError: SecurityError: Failed to read the 'localStorage' property from 'Window': Access is denied for this document.
at app/anvil_extras/storage.py, line 103
  called from StorageDemo, line 5
  called from Main, line 9
```
I suppose it's something to do with the iframe? This was not happening just 24 hours ago, but now it persistently is.

Clone link: https://anvil.works/build#clone:X32NITRHDRRHDOFO=OIDA3G3GQVMLMMT2LPC7BV3O

**Comment by s-cork** [comment]
I'll have a look a bit later. Which browser? if your browser was  recently updated this might also have caused the new behaviour.

If you're working on a demo for local storage it's always good to make an issue first so that we don't double up on work. You never know - someone might already have written a demo 😉.

**Comment by hugetim** [comment]
Right, I'll check next time on the demo, though I partly do those for my own learning. 🙂

I was using Chrome, Version 91.0.4472.124. It could have updated since last time, though I'm not sure.

**Comment by s-cork** [comment]
Ok so seems related to this issue: https://stackoverflow.com/questions/30481516/iframe-in-chrome-error-failed-to-read-localstorage-from-window-access-deni
Try checking your settings for cookies. chrome://settings/content/cookies
I can reproduce if I change those settings.

If that works we can think about some solutions/workarounds/doc updates.

**Comment by hugetim** [comment]
Yep, it's the "Block third-party cookies" setting (my company's default)
that does it.

**Comment by s-cork** [comment]
solutions then:
1. try catch - and use a dictionary if localStorage is not available. print a warning in this case
2. try catch - raise the exception and point to the solution

Something else?

Then add documentation about using `local_storage` inside an iframe.

**Comment by s-cork** [comment]
@hugetim did you want to be assigned this issue? And or have an opinion on the best resolution?

**Comment by hugetim** [comment]
I like your proposed #1, with the warning also pointing to the solution. I
think I can take it on, if that would be helpful.

**Comment by hugetim** [comment]
Thinking about how I'd want to handle this in an app, I think I'd like to add the warning to the API somehow, possibly a boolean named something like `access_denied`? Then in my app, I could have something like:

```python
if storage.access_denied:
    anvil.alert("Due to your browser settings, any changes will be lost after closing this browser session.")
```

**Comment by s-cork** [comment]
Yeah I like that idea. I think we should make this a functions call. `check_available()` or something.

Maybe there's a way to keep the error object as well. `get_access_info()` which could be similar to a json return value. So a dictionary with status and message as keys. Where the message is the error message if our status is `"fail"`. 

Then if we still want a fallback in a backwards compatible way we can add a function to do so. But we probably don't need that for now.

**Comment by hugetim** [comment]
@s-cork, please check out the discussion on my initial pull request here: https://github.com/anvilistas/anvil-extras/pull/129  Seems to be moving in a slightly different direction. Could still use this `get_access_info()` idea, though.

At the moment, the sticking point for my proposing a revised pull request is that I can't figure out the proper way to catch the "ExternalError" exception.

**Comment by s-cork** [comment]
You can't catch an ExternalError - But it's base class is Exception so you can just catch the Exception.
And if you want to know it's an ExternalError you can see if the `repr(e).startswith('External')` or something

**Comment by hugetim** [comment]
Is there a good reason ExternalError isn't exposed, or would it make sense for me to post an Anvil Feature Request to expose it?

**Comment by s-cork** [comment]
It's not part of the python builtins so you wouldn't expect to exist in all namespaces like Exception does. 

Which means it should be imported. But it's actually part of skulpt internals rather than an anvil specific thing. 

It's showing up more now because of the recent anvil.js updates. And if it had to live somewhere it would probably live in anvil.js. 

TLDR- yes a feature request would be a good idea.
